src/System.IO.Pipelines/ReadableBuffer.cs

```diff
@@ -33,7 +31,7 @@ public struct ReadableBuffer : ISequence<ReadOnlyMemory<byte>>
         /// <summary>
         /// Determines if the <see cref="ReadableBuffer"/> is empty.
         /// </summary>
-        public bool IsEmpty => Length == 0;
+        public bool IsEmpty => _first.IsEmpty && Length == 0;
```

Comment:
I'm wondering if it's better to optimize even in the case when _first is empty. I'm a little worried that there might be a pathological (but still reasonably plausible) scenario where the consumer will read till the end of a segment of Memory and then check IsEmpty repeatedly on new Readable buffers. This could be particularly bad if the consumer falls far enough behind the producer.

MemoryPoolIterator.IsEnd short-circuits aggressively.